### PR TITLE
[SPARK-7425] [ML] spark.ml Predictor should support other numeric types for label

### PR DESCRIPTION
**Pull request by BenFradet**
Currently, the Predictor abstraction expects the input labelCol type to be DoubleType, but we should support other numeric types. This will involve updating the PredictorParams.validateAndTransformSchema method.

### CODE REVIEW

**Comment by BenFradet** [comment]
This is still a wip but remarks are welcome.

**Comment by sethah** [comment]
I assume since this is a WIP you are still going to add test cases for the other predictors? Additionally, since ShortType and DecimalType also extend NumericType, I think the tests should include those cases.

File: mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala [hunk 1]
```diff
@@ -275,6 +274,40 @@ class DecisionTreeClassifierSuite extends SparkFunSuite with MLlibTestSparkConte
     val model = dt.fit(df)
   }
 
+  test("DecisionTree should support all NumericType labels") {
+    val dfWithIntLabels = TreeTests.setMetadata(sqlContext.createDataFrame(Seq(
+      (0, Vectors.dense(0, 2, 3)),
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala]: It might be less verbose to create the dataframe once and then add different types of label columns. Something like:

File: mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala [hunk 2]
```diff
@@ -275,6 +274,40 @@ class DecisionTreeClassifierSuite extends SparkFunSuite with MLlibTestSparkConte
     val model = dt.fit(df)
   }
 
+  test("DecisionTree should support all NumericType labels") {
+    val dfWithIntLabels = TreeTests.setMetadata(sqlContext.createDataFrame(Seq(
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala]: It might be less verbose to create the dataframe once, and then add the other label column types to the same data frame. Something like:

```scala
val dfWithTypes = df
      .withColumn("shortLabel", df("labelIndex").cast(ShortType))
      .withColumn("longLabel", df("labelIndex").cast(LongType))
      .withColumn("intLabel", df("labelIndex").cast(IntegerType))
      .withColumn("floatLabel", df("labelIndex").cast(FloatType))
      .withColumn("decimalLabel", df("labelIndex").cast(DecimalType(10, 0)))
```

Then just change the label column between training. I'm not sure which way is better, but this would reduce copying the code ~5 times per test.

**Comment by BenFradet** [comment]
@sethah Yup, I plan to make it exhaustive but it'll take a bit of time.

File: mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala (hunk 2 above)
Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala]: I think so too, thanks.

File: mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala [hunk 3]
```diff
@@ -275,6 +274,40 @@ class DecisionTreeClassifierSuite extends SparkFunSuite with MLlibTestSparkConte
...
+    )).toDF("label", "features"), 2)
+
+    val dt = new DecisionTreeClassifier()
+      .setLabelCol("label")
+      .setFeaturesCol("features")
+
+    dt.fit(dfWithIntLabels)
+    dt.fit(dfWithFloatLabels)
+    dt.fit(dfWithLongLabels)
+  }
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala]: I'm not sure it's not clear what this test is checking for. It looks like it just checks that no errors are thrown during training the tree. Maybe we should train a tree with a data frame with `DoubleType` as the column and check equality between training trees with other label column types?

```scala
val doubleModel = dt.fit(dfWithDoubleLabels)
val intModel = dt.fit(dfWithIntLabels)
TreeTests.checkEqual(doubleModel, intModel)
```

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala [hunk 4]
```diff
@@ -100,6 +101,40 @@ class GBTClassifierSuite extends SparkFunSuite with MLlibTestSparkContext {
     Utils.deleteRecursively(tempDir)
   }
 
+  test("GradientBoostedTrees should support all NumericType labels") {
+    val dfWithIntLabels = TreeTests.setMetadata(sqlContext.createDataFrame(Seq(
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: I'm wondering if there is some way to reuse the code here since we need to implement roughly the same test (train with all different label column types, check that they are the same as training with double type) for each different predictor. Maybe adding something in a helper function under `MLTestingUtils`?

**Comment by BenFradet** [comment]
I noticed [AFTSurvivalRegression](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala#L99) uses its own `validateAndTransformSchema` method.

Should I make it support other numeric types in this PR or in another one?

File: mllib/src/test/scala/org/apache/spark/ml/classification/OneVsRestSuite.scala [hunk 5]
```diff
@@ -160,6 +160,59 @@ class OneVsRestSuite extends SparkFunSuite with MLlibTestSparkContext {
...
+    )).toDF("label", "features")
+
+    val ovr = new OneVsRest()
+      .setClassifier(new LogisticRegression)
+      .setLabelCol("label")
+      .setFeaturesCol("features")
+
+    // thrown by AttributeFactory#fromStructField and not by Predictor#validateAndTransformSchema
+    // because OneVsRest reimplements the fit method
+    intercept[IllegalArgumentException] {
```
Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/classification/OneVsRestSuite.scala]: This is caused by [AttributeFactory#fromStructField](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/attribute/attributes.scala#L130).

I don't know what's the best way to handle this, input welcome.

**Comment by BenFradet** [comment]
cc @jkbradley

**Comment by BenFradet** [comment]
Also, what about the other things expecting a `labelCol` of `DoubleType` like the different evaluators?
Should this be handled in a follow-up jira/pr?

**Comment by BenFradet** [comment]
Failing tests do not seem related to the changes introduced, triggering a new build.
`org.apache.spark.sql.hive.MetastoreDataSourcesSuite.insert into a table`
`org.apache.spark.sql.hive.MetastoreDataSourcesSuite.SPARK-8156:create table to specific database by 'use dbname'`

**Comment by BenFradet** [comment]
Jenkins, retest this please

**Comment by BenFradet** [comment]
Pinging @jkbradley and @mengxr

**Comment by BenFradet** [comment]
Pinging @jkbradley and @mengxr

**Comment by BenFradet** [comment]
@mengxr @jkbradley is there any interest in this pr or should i close it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala [hunk 6]
```diff
@@ -151,7 +151,7 @@ abstract class NumericType extends AtomicType {
 }
 
 
-private[sql] object NumericType extends AbstractDataType {
+private[spark] object NumericType extends AbstractDataType {
```
Comment (MLnick) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: Why is this change required?

**Comment by MLnick** [comment]
@BenFradet I'll review - there seem to be merge conflicts that need resolving.

**Comment by BenFradet** [comment]
Great! Thanks.
I'll fix those conflicts later today.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala (hunk 6 above)
Comment (BenFradet) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: Because it's now used in [Predictor.scala]{https://github.com/apache/spark/pull/10355/files#diff-5fe763c1d5d2d36c0154204028cd3585R56}

Comment (MLnick) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: I don't think it's required - see https://github.com/apache/spark/pull/9777/files#diff-a364fbaaec2088b49d5d1ceaea4a1c41R73 where OneHotEncoder uses `isInstanceOf[NumericType]` without making this object `private[spark]`. Here `NumericType` refers to the abstract class [here](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala#L144).

**Comment by BenFradet** [comment]
@MLnick you were right, thanks for spotting that.

**Comment by MLnick** [comment]
@BenFradet if it's not too much trouble, I think you can do `AFTSurvivalRegression` and `IsotonicRegression` here.

**Comment by BenFradet** [comment]
Surething

**Comment by MLnick** [comment]
@BenFradet I'm ok with doing the other things like evaluators in a different PR - however the only concern I'd have is it would temporarily break things like `CrossValidator` that use evaluators and a non-Double `labelCol` for the model. It's ok if you're able to do the follow up soon. Otherwise it may be worth thinking about just doing them here.

File: mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala [hunk 7]
```diff
@@ -334,6 +334,51 @@ class DecisionTreeClassifierSuite extends SparkFunSuite with MLlibTestSparkConte
     assert(importances.toArray.forall(_ >= 0.0))
   }
 
+  test("should support all NumericType labels") {
```
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala]: It's not a major issue, but there is a large amount of code duplication in all the tests. I know you've already tried to pare it down, but if it's possible to further reduce it by extracting out some methods, that would be cleaner.

Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala]: The code duplication is mostly related to the test datasets.
What I can do is reduce their size to one or two observations, what do you think?

Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala]: I was thinking more like create a util method (or use an existing one if there is one) to generate the required test datasets. It's not a big deal though, just a "nice to have"

Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/classification/DecisionTreeClassifierSuite.scala]: OK, will do.

**Comment by BenFradet** [comment]
Here's what's left to do:
- IsotonicRegression
- GeneralizedLinearRegression
- AFTSurvival

Otherwise, I'd prefer doing the evaluators and other estimators (like rformula and chiSqSelector) in other prs

File: mllib/src/main/scala/org/apache/spark/ml/Predictor.scala [hunk 8]
```diff
@@ -49,10 +50,7 @@ private[ml] trait PredictorParams extends Params
     validateParams()
     // TODO: Support casting Array[Double] and Array[Float] to Vector when FeaturesType = Vector
     SchemaUtils.checkColumnType(schema, $(featuresCol), featuresDataType)
-    if (fitting) {
-      // TODO: Allow other numeric types
-      SchemaUtils.checkColumnType(schema, $(labelCol), DoubleType)
-    }
+    if (fitting) SchemaUtils.checkNumericType(schema, $(labelCol))
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/Predictor.scala]: Spark style prefers to keep the braces format for if statements. Can you change it back please?

File: mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala [hunk 9]
```diff
@@ -347,6 +348,34 @@ class AFTSurvivalRegressionSuite
     }
   }
 
+  test("should support all NumericType labels") {
+    val dfs = MLTestingUtils.genRegressionDFWithNumericLabelCol(sqlContext, "label", "features")
```
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala]: I think you'll need to adjust the generated data here as it requires an additional 'censor' col.

**Comment by MLnick** [comment]
@BenFradet I linked 2 additional sub-tasks to the [umbrella JIRA](https://issues.apache.org/jira/browse/SPARK-11107)

File: mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala (hunk 9 above)
Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala]: Thanks for the insight, I'll fix this later today.

**Comment by BenFradet** [comment]
@MLnick yup thanks, I'll get to those once I'm done with this one.

**Comment by MLnick** [comment]
Looks fine to me - latest test result indicates there may be merge conflicts?

I'd like @jkbradley to take a quick pass too.

**Comment by MLnick** [comment]
@sethah could you also take a final pass?

**Comment by BenFradet** [comment]
Yup, I'll rebase tonight and start working on the related jiras.

File: mllib/src/test/scala/org/apache/spark/ml/classification/RandomForestClassifierSuite.scala [hunk 10]
```diff
@@ -178,6 +179,34 @@ class RandomForestClassifierSuite extends SparkFunSuite with MLlibTestSparkConte
...
+
+    val expected = rf.setLabelCol("label")
+      .fit(TreeTests.setMetadata(dfs(DoubleType), 2, "label"))
+    dfs.keys.filter(_ != DoubleType).foreach { t =>
+      TreeTests.checkEqual(expected,
+        rf.setLabelCol("label").fit(TreeTests.setMetadata(dfs(t), 2, "label")))
+    }
+  }
+
+  test("shouldn't support non NumericType labels") {
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/classification/RandomForestClassifierSuite.scala]: I still think the code duplication is a big concern. It might not be that hard to add some utility function that can clean this up. For example I was able to get this working:

In _MLTestingUtils.scala_
```scala
def checkRejectsNonNumericType(est: Predictor[_, _, _], sqlContext: SQLContext) = {
    val dfWithStringLabels =
      MLTestingUtils.generateDFWithStringLabelCol(sqlContext, est.getLabelCol, est.getFeaturesCol)

    val thrown = intercept[IllegalArgumentException] {
      est.fit(dfWithStringLabels)
    }
    assert(thrown.getMessage contains
      "Column label must be of type NumericType but was actually of type StringType")
  }
```

Then in the actual test suites:
```scala
test("shouldn't support non NumericType labels") {
    MLTestingUtils.checkRejectsNonNumericType(new RandomForestClassifier, sqlContext)
}
```
There will be some complication for some of the estimators like OneVsRest and IsotonicRegression (why doesn't it extend Predictor?) but I think it will be worth it to get this right since future estimators will have to implement this as well. It would be really nice to have something like we do for params where there is just a one-liner `ParamsSuite.checkParams(new RandomForestClassifier)`. For testing that all numeric types are supported, we could have a utility method that produces all actual and expected results, then check for equality inside the individual test suites, like:

```scala
val models = MLTestingUtils.fitAllNumericTypes(new NaiveBayes, sqlContext)
    models.foreach { case (expected, actual) =>
      assert(expected.pi === actual.pi)
      assert(expected.theta === actual.theta)
    }
```

I'm open to feedback on this, realizing it could be hard to generalize this for every case.

File: mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala [hunk 11]
```diff
@@ -27,4 +31,56 @@ object MLTestingUtils {
     assert(copied.parent.uid == model.parent.uid)
     assert(copied.parent == model.parent)
   }
+
+  def genClassifDFWithNumericLabelCol(
+    sqlContext: SQLContext,
+    labelColName: String,
+    featuresColName: String
+  ): Map[NumericType, DataFrame] = {
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala]: nit: Spark style seems to be to put the return type on the line above, if it will fit. So:

```scala
def genClassifDFWithNumericLabelCol(
    sqlContext: SQLContext,
    labelColName: String,
    featuresColName: String): Map[NumericType, DataFrame] = {
```
[Example](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala#L168)

Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala]: I'll modify the whole file then.

File: mllib/src/test/scala/org/apache/spark/ml/classification/RandomForestClassifierSuite.scala (hunk 10 above)
Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/classification/RandomForestClassifierSuite.scala]: Yup, good idea, I'll investigate.

Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/classification/RandomForestClassifierSuite.scala]: I wanted to be able to do something like:

```scala
def checkNumericTypes[T <: Regressor, M <: Model](
    regressor: T, sqlContext: SQLContext)(asserts: Seq[(M, M) => Unit]): Unit = {
  val dfs = MLTestingUtils.genRegressionDFWithNumericLabelCol(sqlContext, "label", "features")
    
  val expected = regressor.fit(dfs(DoubleType))
  dfs.keys.filter(_ != DoubleType).foreach { t =>
    val actual = regressor.fit(dfs(t))
    asserts.foreach(f => f(expected, actual))
  }
}
```

But unfortunately that doesn't work since I'm not able to capture which model is being created by the regressor's `fit` method.

Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/classification/RandomForestClassifierSuite.scala]: Here's what I managed to make work:

```scala
  def checkNumericTypes[M <: RegressionModel[_, M], T <: Regressor[_, _, M]](
      regressor: T, sqlContext: SQLContext)(check: (M, M) => Unit): Unit = {
    val dfs = MLTestingUtils.genRegressionDFWithNumericLabelCol(sqlContext, "label", "features")
    
    val expected = regressor.fit(dfs(DoubleType))
    val actuals = dfs.keys.filter(_ != DoubleType).map(t => regressor.fit(dfs(t)))
    actuals.foreach(actual => check(expected, actual))
  }
```

which could be used like:

```scala
MLTestingUtils.checkNumericTypes[LinearRegressionModel, LinearRegression](
    new LinearRegression(), sqlContext) { (expected, actual) =>
      assert(expected.intercept === actual.intercept)
      assert(expected.coefficients === actual.coefficients)
    }
```

**Comment by jkbradley** [comment]
Sorry for the long delay!  I'll take a look now

**Comment by jkbradley** [comment]
@BenFradet Are you actively working on this?  if so, I'll wait to review it.

File: mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala [hunk 12]
```diff
@@ -183,12 +183,12 @@ class AFTSurvivalRegression @Since("1.6.0") (@Since("1.6.0") override val uid: S
    * Extract [[featuresCol]], [[labelCol]] and [[censorCol]] from input dataset,
    * and put it in an RDD with strong types.
    */
-  protected[ml] def extractAFTPoints(dataset: DataFrame): RDD[AFTPoint] = {
-    dataset.select($(featuresCol), $(labelCol), $(censorCol)).rdd.map {
-      case Row(features: Vector, label: Double, censor: Double) =>
-        AFTPoint(features, label, censor)
-    }
-  }
+  protected[ml] def extractAFTPoints(dataset: DataFrame): RDD[AFTPoint] =
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala]: I prefer to keep braces around multiline methods like this.

**Comment by BenFradet** [comment]
@jkbradley yes, I'm currently trying to incorporate @sethah's comments and keep code duplication to a strict minimum.

**Comment by BenFradet** [comment]
@jkbradley @sethah @MLnick I think it's in a pretty good state now, what do you guys think?

**Comment by jkbradley** [comment]
I'll take a look

File: mllib/src/main/scala/org/apache/spark/ml/util/SchemaUtils.scala [hunk 13]
```diff
@@ -61,6 +61,20 @@ private[spark] object SchemaUtils {
   }
 
   /**
+    * Check whether the given schema contains a column of the numeric data type.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/util/SchemaUtils.scala]: indent 1 space less (can update IntelliJ code style settings: Editor -> Code Style -> Scala -> ScalaDoc tab -> uncheck Use scaladoc indent for leading asterisk

File: mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala [hunk 14]
```diff
@@ -17,14 +17,116 @@
...
+      (3, Vectors.dense(3)),
+      (4, Vectors.dense(4))
+    )).toDF(labelColName, featuresColName)
+
+    val types =
+      Seq(ShortType, LongType, IntegerType, FloatType, ByteType, DoubleType, DecimalType(10, 0))
+    types
+      .map(t => t -> df.select(col(labelColName).cast(t), col(featuresColName)))
+      .map { case (t, d) =>
+        t -> TreeTests.setMetadata(d, 2, labelColName).withColumn(censorColName, lit(0.0))
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala]: setMetadata: classes should be 0, not 2

**Comment by jkbradley** [comment]
This looks to be in very good shape.  My main comment is about unit test time: the tests take a significant amount of time.  Wherever it is easy to do, could you set the algorithm parameters to make for faster tests?  E.g., use 1 iteration for LogisticRegression, use maxDepth 0 for trees, etc.

**Comment by BenFradet** [comment]
Sure, thanks for your review.

File: mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala [hunk 15]
```diff
@@ -17,14 +17,116 @@
...
-object MLTestingUtils {
+object MLTestingUtils extends SparkFunSuite {
   def checkCopy(model: Model[_]): Unit = {
     val copied = model.copy(ParamMap.empty)
       .asInstanceOf[Model[_]]
     assert(copied.parent.uid == model.parent.uid)
     assert(copied.parent == model.parent)
   }
+
+  def checkPredictorAcceptAllNumericTypes[M <: PredictionModel[_, M], T <: Predictor[_, _, M]](
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala]: Is there a reason to have two separate methods for predictor and estimator? I also tend to prefer _not_ separating the "accept" and "reject" checks, and instead have them in a single function call. This will reduce the complexity of having to decide which one to use (and why) and also reduce the number of function calls in future tests. I was able to get the tests working having a single method for accept/reject and estimator/predictor, like this:

```scala
  def checkEstimatorNumericTypes[M <: Model[M], T <: Estimator[M]](
      predictor: T,
      isClassification: Boolean,
      sqlContext: SQLContext)(check: (M, M) => Unit): Unit = {
    val dfs = if (isClassification) {
      genClassifDFWithNumericLabelCol(sqlContext)
    } else {
      genRegressionDFWithNumericLabelCol(sqlContext)
    }
    val expected = predictor.fit(dfs(DoubleType))
    val actuals = dfs.keys.filter(_ != DoubleType).map(t => predictor.fit(dfs(t)))
    actuals.foreach(actual => check(expected, actual))

    val dfWithStringLabels = generateDFWithStringLabelCol(sqlContext)
    val thrown = intercept[IllegalArgumentException] {
      predictor.fit(dfWithStringLabels)
    }
    assert(thrown.getMessage contains
      "Column label must be of type NumericType but was actually of type StringType")
  }
```

I had to change the fit method of OneVsRest to include a call to `transformSchema(dataset.schema)` so that it would fail properly on non-numeric label column. I think it's better to change the fit method so that all ML estimators provide consistent errors. Also note that the `isClassification` flag is needed because MLPClassifier does not inherit from Classifier, and so it is not easy to infer that it is a classification estimator.

Comment (BenFradet) [inline mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala]: good point, will update

File: mllib/src/main/scala/org/apache/spark/ml/Predictor.scala [hunk 16]
```diff
@@ -26,7 +26,7 @@ import org.apache.spark.mllib.regression.LabeledPoint
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.{DataFrame, Row}
 import org.apache.spark.sql.functions._
-import org.apache.spark.sql.types.{DataType, DoubleType, StructType}
+import org.apache.spark.sql.types.{DataType, DoubleType, NumericType, StructType}
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/Predictor.scala]: minor, but I believe the `NumericType` import is no longer necessary

**Comment by MLnick** [comment]
One very minor comment on imports, and think there are now merge conflicts with latest master that need fixing, subject to those LGTM.

**Comment by BenFradet** [comment]
Will fix that, thanks.

**Comment by jkbradley** [comment]
LGTM
I'll go ahead and merge this with master before a conflict happens
Thanks for the PR!

**Comment by BenFradet** [comment]
No problem, thanks for reviewing.
